week4/week4.playground/Contents.swift

```diff
@@ -0,0 +1,203 @@
+
+- Create a function called sumOfArray that takes a parameter of [Int?] and returns an Int. The function should return the sum of the array values passed that are not nil. Call the function passing the numsWithNil array, and print out the results.
+
+- Create a function called commonElementsSet that takes two parameters of [Int] and returns a Set<Int> (set of Int). The function will return a Set<Int> of the values in both arrays.
+
+- Call the function commonElementsSet passing the arrays numsBy2, numsBy4, and print out the results.
+*/
+
+let nums = Array(0...100)
+let numsWithNil: [Int?] = [79, nil, 80, nil, 90, nil, 100, 72]
+
+var numsBy2 = [Int]()
+for number in stride(from: 2, to: 100, by: 2) {
+  numsBy2.append(number)
+}
+print("numsBy2 array will be \(numsBy2)")
+
+var numsBy4 = [Int]()
+for number in stride(from: 2, to: 100, by: 4) {
+  numsBy4.append(number)
+}
+print("numsBy4 array will be \(numsBy4)")
+
+func evenNumbersArray(numberArray: [Int]) -> [Int] {
+  var evenNumbers = numberArray.filter { $0 % 2 == 0 }
+  return evenNumbers
+}
+
+print(evenNumbersArray(numberArray: nums))
+
+func sumOfArray(optionalArray: [Int?]) -> Int {
+  var sumOfNonNilValues = optionalArray.compactMap { Int($0 ?? 0)}.reduce(0, +)
+  return sumOfNonNilValues
+}
+
+print("sum of non nil values is \(sumOfArray(optionalArray: numsWithNil))")
+
+func commonElementsSet(numArrayOne: [Int], numArrayTwo: [Int]) -> Set<Int> {
+  let combinedArray = numArrayOne + numArrayTwo
+  let resultSet = Set(combinedArray)
+  return resultSet
+}
+
+print("Set of two arrays \(commonElementsSet(numArrayOne: numsBy2, numArrayTwo: numsBy4))")
+
+
+// d) Create a struct called Square that has a stored property called sideLength and a computed property called area. Create an instance of Square and print the area.
+
+struct Square {
+  var sideLength: Int
+  var area: Int {
+    sideLength * sideLength
+  }
+}
+
+let mySquare = Square(sideLength: 10)
+
+print("the area of mySquare is \(mySquare.area)")
+
+// Above and beyond
+protocol Shape {
+  func calculateArea() -> Double
+}
+
+struct Circle: Shape {
+  var radius: Double
+  private var pi = 3.14
+  
+  init(radius: Double) {
+    self.radius = radius
+  }
+  
+  func calculateArea() -> Double {
+    let area = pi * radius * radius
+    return area
+  }
+}
+
+struct Rectangle: Shape {
+  var length: Double
+  var width: Double
+  func calculateArea() -> Double {
+    let area = length * width
+    return area
+  }
+}
+
+let myCircle = Circle(radius: 10)
+print("Area of circle is \(myCircle.calculateArea())")
+
+let myrectangle = Rectangle(length: 10, width: 10)
+print("Area of rectangle is \(myrectangle.calculateArea())")
+
+protocol VolumeOfAShape: Shape {
+  func calculateVolume() -> Double
```

Comment:
This is unfortunately not the correct solution for the task. It's definitely a one way to do it, but the task was to create an `extension` on the `protocol Shape`, not another `protocol`. 
Like this:
```
extension Shape {
    func calculateVolume() -> Double
}
```

Then you'd get an error and see that protocol extensions require a default implementation so you had to add that as well.